File: server/java/omdb/src/main/java/com/metaobjects/manager/db/migrate/SqlType.java
package com.metaobjects.manager.db.migrate;

/**
 * Canonical, dialect-neutral SQL type — the cross-language spine (port of migrate-ts SqlType).
 * Both the expected-snapshot builder (metadata->type) and the introspector (db->type) produce this;
 * the differ compares canonical-to-canonical (record equals == sqlTypeEquals); emit re-renders to
 * dialect SQL.
 *
 * Per spec §5.2.
 */
public sealed interface SqlType
        permits SqlType.Text, SqlType.Int, SqlType.Real, SqlType.Numeric, SqlType.Bool,
                SqlType.Timestamp, SqlType.Date, SqlType.Json, SqlType.Blob, SqlType.Uuid {

    /** maxLength == null means unbounded (TEXT). */
    record Text(Integer maxLength) implements SqlType {}

    /** bits: 32 (INTEGER) or 64 (BIGINT). Named Int to avoid shadowing java.lang.Integer. */
    record Int(int bits) implements SqlType {}

    record Real() implements SqlType {}

    record Numeric(Integer precision, Integer scale) implements SqlType {}

    record Bool() implements SqlType {}

    record Timestamp(boolean withTimezone) implements SqlType {}

    record Date() implements SqlType {}

    record Json() implements SqlType {}

    record Blob() implements SqlType {}

    record Uuid() implements SqlType {}

    /**
     * Returns true if changing column type from {@code from} to {@code to} is provably non-lossy.
     * Port of migrate-ts {@code isWidening}. Conservative on purpose — false negatives just mean
     * the user has to explicitly allow the change; false positives could silently corrupt data.
     *
     * Returns false for identical types (caller should not emit a change-column-type in that case;
     * this is defensive).
     *
     * Per spec §6.5.
     */
    static boolean isWidening(SqlType from, SqlType to) {
        if (from.equals(to)) return false;
        if (from.getClass() != to.getClass()) return false;        // cross-kind: lossy

        if (from instanceof Text f && to instanceof Text t) {
            if (f.maxLength() == null) return false;               // unbounded -> bounded: lossy
            if (t.maxLength() == null) return true;                // bounded -> unbounded: widening
            return t.maxLength() >= f.maxLength();                 // bounded both: widening iff new >= old
        }
        if (from instanceof Int f && to instanceof Int t) {
            return t.bits() >= f.bits();
        }
        if (from instanceof Numeric f && to instanceof Numeric t) {
            int fp = f.precision() == null ? 0 : f.precision();
            int fs = f.scale()     == null ? 0 : f.scale();
            int tp = t.precision() == null ? 0 : t.precision();
            int ts = t.scale()     == null ? 0 : t.scale();
            // widening iff precision expands, scale unchanged, and integer-part expands
            return tp >= fp && ts == fs && (tp - ts) >= (fp - fs);
        }

        // Real/Bool/Date/Json/Blob/Uuid/Timestamp: any same-kind difference is not widening
        return false;
    }
}

File: server/typescript/packages/migrate-ts/src/sql-type.ts
/**
 * Canonical, dialect-neutral SQL type. Both buildExpectedSchema (metadata→snapshot)
 * and introspect (db→snapshot) produce this shape; diff compares canonical to
 * canonical; emit re-renders to dialect-specific SQL.
 *
 * Per spec §5.2.
 */
export type SqlType =
  | { kind: "text"; maxLength?: number }
  | { kind: "integer"; bits: 32 | 64 }
  | { kind: "real" }        // DOUBLE PRECISION (float8) — field.double
  | { kind: "real4" }       // REAL (float4, single precision) — field.float
  | { kind: "numeric"; precision?: number; scale?: number }
  | { kind: "boolean" }
  | { kind: "timestamp"; withTimezone: boolean }
  | { kind: "date" }
  | { kind: "time" }
  | { kind: "json" }
  | { kind: "blob" }
  | { kind: "uuid" }
  | { kind: "inet" }        // Postgres-native inet (IPv4/IPv6) — field.inet (ADR-0036/0037 Wave 3)
  | { kind: "array"; element: SqlType };   // native SQL array (e.g. uuid[], text[])

/** Structural equality on SqlType. */
export function sqlTypeEquals(a: SqlType, b: SqlType): boolean {
  if (a.kind !== b.kind) return false;
  switch (a.kind) {
    case "text":
      return a.maxLength === (b as Extract<SqlType, { kind: "text" }>).maxLength;
    case "integer":
      return a.bits === (b as Extract<SqlType, { kind: "integer" }>).bits;
    case "numeric": {
      const bn = b as Extract<SqlType, { kind: "numeric" }>;
      return a.precision === bn.precision && a.scale === bn.scale;
    }
    case "timestamp":
      return a.withTimezone === (b as Extract<SqlType, { kind: "timestamp" }>).withTimezone;
    case "array":
      return sqlTypeEquals(a.element, (b as Extract<SqlType, { kind: "array" }>).element);
    case "real":
    case "real4":
    case "boolean":
    case "date":
    case "time":
    case "json":
    case "blob":
    case "uuid":
    case "inet":
      return true;
  }
}

/**
 * Returns true if changing column type from `from` to `to` is provably non-lossy.
 * Per spec §6.5. Conservative on purpose — false negatives just mean the user has
 * to pass `allow.typeChange`; false positives could silently corrupt data.
 *
 * Returns false for identical types (caller should not emit a change-column-type
 * in that case at all; this is defensive).
 */
export function isWidening(from: SqlType, to: SqlType): boolean {
  if (sqlTypeEquals(from, to)) return false;
  if (from.kind !== to.kind) return false;       // cross-kind: always lossy

  switch (from.kind) {
    case "text": {
      const t = to as Extract<SqlType, { kind: "text" }>;
      // unbounded → bounded: lossy
      if (from.maxLength === undefined) return false;
      // bounded → unbounded: widening
      if (t.maxLength === undefined) return true;
      // bounded both: widening iff new ≥ old
      return t.maxLength >= from.maxLength;
    }
    case "integer": {
      const t = to as Extract<SqlType, { kind: "integer" }>;
      return t.bits >= from.bits;
    }
    case "numeric": {
      const t = to as Extract<SqlType, { kind: "numeric" }>;
      const fp = from.precision ?? 0;
      const fs = from.scale ?? 0;
      const tp = t.precision ?? 0;
      const ts = t.scale ?? 0;
      // widening iff p2 ≥ p1 AND s2 = s1 AND (p2 - s2) ≥ (p1 - s1)
      return tp >= fp && ts === fs && (tp - ts) >= (fp - fs);
    }
    // real/real4/boolean/date/json/blob/uuid: same kind already handled by sqlTypeEquals;
    // any difference here means the discriminant matched but structural equality failed
    // (impossible given SqlType has no other variants for these kinds). Defensive false.
    case "real":
    case "real4":
    case "boolean":
    case "date":
    case "time":
    case "json":
    case "blob":
    case "uuid":
    case "inet":
    case "timestamp":
    case "array":
      // Array element-type changes are not auto-widened — require explicit allow.
      return false;
  }
}
